pkg/store/bucket.go

```diff
@@ -1030,6 +1032,19 @@ func (s *bucketBlockSet) getFor(mint, maxt, minResolution int64) (bs []*bucketBl
 	if len(bs) == 0 {
 		return s.getFor(mint, maxt, s.resolutions[i])
 	}
+
+	until := len(bs) - 1
+	for j := 0; j < until; j++ {
```

Comment:
It's tricky, need to dive more but generally makes sense. Explaining what we are seeing now. Thanks for tests catching this as well! 

All of this assumes that blocks within different resolutions are aligned ideally right? Can we comment this somehow?

Also be aware of this: https://github.com/improbable-eng/thanos/pull/1104 Hope you algorthim would be nicely extendable I guess.
 cc @mjd95

Reply:
They always should be aligned ideally or on the same timestamps because in other cases we would get a problem known as "overlapping blocks".

This change should be easy to integrate into that PR.

Reply:
OK, It makes total sense BUT to me `left := s.getFor(mint, bs[0].meta.MinTime, s.resolutions[i])` and `right` are just `between` 0 and first min and lastMax and maxt, right?

Can we remove this left and right and treat is properly in this loop as `between`? I think it would simplify flow and algorithm.

Reply:
The idea sounds and I played around with it a bit but I doubt that this function can be made more succinct because we have three distinct cases of the `append` call: at the beginning of the array, between two elements, and at the end. Please correct me if I'm wrong.

Reply:
Hm, is it really too complex? I will propose something to your PR

Reply:
Proposed: https://github.com/GiedriusS/thanos/pull/1

IMO it's way simpler, what do you think? @GiedriusS 

```go
// getFor returns a time-ordered list of blocks that cover date between mint and maxt.
// Blocks with the biggest resolution possible but not bigger than the given max resolution are returned.
func (s *bucketBlockSet) getFor(mint, maxt, maxResolutionMillis int64) (bs []*bucketBlock) {
	if mint == maxt {
		return nil
	}

	s.mtx.RLock()
	defer s.mtx.RUnlock()

	// Find first matching resolution.
	i := 0
	for ; i < len(s.resolutions) && s.resolutions[i] > maxResolutionMillis; i++ {
	}

	// Fill the given interval with the blocks for the current resolution.
	// Our current resolution might not cover all data, so recursively fill the gaps with higher resolution blocks if there is any.
	start := mint
	for _, b := range s.blocks[i] {
		if b.meta.MaxTime <= mint {
			continue
		}
		if b.meta.MinTime >= maxt {
			break
		}

		if i+1 < len(s.resolutions) {
			bs = append(bs, s.getFor(start, b.meta.MinTime, s.resolutions[i+1])...)
		}
		bs = append(bs, b)
		start = b.meta.MaxTime
	}

	if i+1 < len(s.resolutions) {
		bs = append(bs, s.getFor(start, maxt, s.resolutions[i+1])...)
	}
	return bs
}
```